Microsoft.Azure.Cosmos/src/Authorization/AzureKeyCredentialAuthorizationTokenProvider.cs

```diff
@@ -0,0 +1,114 @@
+        public override ValueTask AddAuthorizationHeaderAsync(
+                INameValueCollection headersCollection, 
+                Uri requestAddress, 
+                string verb, 
+                AuthorizationTokenType tokenType)
+        {
+            this.CheckAndRefreshTokenProvider();
+            return this.authorizationTokenProvider.AddAuthorizationHeaderAsync(
+                headersCollection, 
+                requestAddress, 
+                verb, 
+                tokenType);
+        }
+
+        public override void Dispose()
+        {
+            if (this.authorizationTokenProvider != null)
+            {
+                this.authorizationTokenProvider.Dispose();
+                this.authorizationTokenProvider = null;
+            }
+        }
+
+        public override ValueTask<(string token, string payload)> GetUserAuthorizationAsync(
+                string resourceAddress, 
+                string resourceType, 
+                string requestVerb, 
+                INameValueCollection headers, 
+                AuthorizationTokenType tokenType)
+        {
+            this.CheckAndRefreshTokenProvider();
+            return this.authorizationTokenProvider.GetUserAuthorizationAsync(
+                    resourceAddress, 
+                    resourceType, 
+                    requestVerb, 
+                    headers, 
+                    tokenType);
+        }
+
+        public override ValueTask<string> GetUserAuthorizationTokenAsync(
+                string resourceAddress, 
+                string resourceType, 
+                string requestVerb, 
+                INameValueCollection headers, 
+                AuthorizationTokenType tokenType, 
+                ITrace trace)
+        {
+            this.CheckAndRefreshTokenProvider();
+            return this.authorizationTokenProvider.GetUserAuthorizationTokenAsync(
+                    resourceAddress, 
+                    resourceType, 
+                    requestVerb, 
+                    headers, 
+                    tokenType, 
+                    trace);
+        }
+
+        public override void TraceUnauthorized(
+                DocumentClientException dce, 
+                string authorizationToken, 
+                string payload)
+        {
+            this.authorizationTokenProvider.TraceUnauthorized(
+                    dce, 
+                    authorizationToken, 
+                    payload);
+        }
+
+        private void CheckAndRefreshTokenProvider()
+        {
+            if (!object.ReferenceEquals(this.currentKeyObject, this.azureKeyCredential.Key))
+            {
+                this.currentKeyObject = this.azureKeyCredential.Key ?? throw new ArgumentNullException($"{nameof(AzureKeyCredential)} has null Key");
+
+                // Credentials changed:
+                // 1. Dispose current token provider
+                // 2. Refresh the provider
+                using (this.authorizationTokenProvider) 
```

Comment:
Perhaps this is a bit more readable?

```suggestion
                this.authorizationTokenProvider?.Dispose();
```